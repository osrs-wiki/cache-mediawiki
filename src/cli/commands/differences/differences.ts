import { Command, Option } from "commander";

import differencesCache from "@/tasks/differences/differences";

const differences = new Command("differences")
  .alias("diffs")
  .description(
    "Generate a page outlining differences between two cache versions."
  )
  .addOption(
    new Option(
      "--indices <indices>",
      "Comma-separated list of index IDs to include in differences (e.g., '2,5,8'). If not provided, all indices are checked."
    )
  )
  .addOption(
    new Option(
      "--ignoreIndices <ignoreIndices>",
      "Comma-separated list of index IDs to exclude from differences (e.g., '2,5,8'). Takes precedence over --indices option."
    )
  )
  .action((options) => {
    differencesCache({
      oldVersion: options.oldCache,
      newVersion: options.newCache,
      method: options.cacheSource,
      type: options.cacheType,
      indices: options.indices,
      ignoreIndices: options.ignoreIndices,
    });
  });

export default differences;
