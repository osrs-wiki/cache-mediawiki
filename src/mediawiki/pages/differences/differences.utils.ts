import {
  MediaWikiContent,
  MediaWikiExternalLink,
  MediaWikiLink,
  type MediaWikiTableRow,
  MediaWikiText,
  MediaWikiBreak,
  MediaWikiHeader,
  MediaWikiTable,
  type MediaWikiTableCell,
} from "@osrs-wiki/mediawiki-builder";
import { diffWords } from "diff";
import _, { flatten } from "underscore";

import {
  IndexFeatures,
  IndexURLType,
  IndexURLs,
  resultNameMap,
} from "./differences.types";
import {
  ArchiveDifferences,
  ChangedResult,
  Difference,
  IndexDifferences,
  Result,
  ResultValue,
} from "../../../tasks/differences/differences.types";

import { jagexHSLtoHex } from "@/utils/colors";
import { capitalize, stripHtmlTags } from "@/utils/string";

/**
 * Format the value a field.
 * @param field The field of a {Result}
 * @param value The unformatted {ResultValue}
 * @returns A formatted string.
 */
export const formatEntryValue = (field: string, value: ResultValue): string => {
  if (
    value === undefined ||
    (value && typeof value === "object" && Object.keys(value).length === 0)
  ) {
    return "";
  }
  if (
    field.toLocaleLowerCase().includes("color") &&
    typeof value === "number"
  ) {
    const hex = jagexHSLtoHex(value);
    return `<span style = "background-color: ${hex};color:white">${value}</span>`;
  } else if (Array.isArray(value)) {
    /*if (field.toLocaleLowerCase().includes("color")) {
      return value.reduce(
        (result, color) =>
          result +
          " " +
          `<span style = "background-color: ${jagexHSLtoHex(
            color
          )};color:white">${color}</span>`,
        ""
      );
    }*/
    const mappedValues = value.map((value) =>
      value == null || value == undefined ? "None" : value
    );
    const stringValue =
      mappedValues.length > 0 ? JSON.stringify(mappedValues) : "";
    return stringValue
      .replaceAll('"None"', "None")
      .replaceAll('"', "")
      .replaceAll(",", ", ");
  } else if (typeof value === "string" || typeof value === "number") {
    return value.toString();
  }
  return JSON.stringify(value)
    .replaceAll('"', "'")
    .replaceAll(",", ", ")
    .replaceAll(":", ": ");
};

/**
 * Format a field's identifier.
 * @param identifier The field identifier key
 * @param value The Result value
 * @param urls Supported urls for names and identifiers
 * @returns
 */
export const formatEntryIdentifier = (
  identifier: string,
  value: ResultValue,
  urls: IndexURLs
): MediaWikiContent[] => {
  switch (identifier) {
    case "id":
      const urlKeys = Object.keys(urls);
      return urlKeys.length > 0
        ? urlKeys.map(
            (url, index) =>
              new MediaWikiExternalLink(
                index === 0 ? `${value}` : `(${index})`,
                `${urls[url as IndexURLType]}${value as string}`
              )
          )
        : [new MediaWikiText(formatEntryValue(identifier, value))];
    case "name":
      return [
        value
          ? new MediaWikiLink(stripHtmlTags(value as string))
          : new MediaWikiText(""),
      ];
    default:
      return [new MediaWikiText(formatEntryValue(identifier, value))];
  }
};

/**
 * Format entry field values and highlight any differences.
 * @param entry The field to check differences for
 * @param field
 * @returns A MediaWikiTableRow with formatted and highlighted columns.
 */
export const getFieldDifferencesRow = (
  entry: ChangedResult,
  field: string
): MediaWikiTableRow => {
  const oldValue = formatEntryValue(field, entry[field].oldValue);
  const newValue = formatEntryValue(field, entry[field].newValue);

  const differences = diffWords(oldValue, newValue);
  let oldValueDiffs = "";
  let newValueDiffs = "";
  differences.forEach((part) => {
    if (!part || !part.value) {
      return;
    }
    if (part.added) {
      newValueDiffs += `<span style="color: #6bc71f">${part.value}</span>`;
    } else if (part.removed) {
      oldValueDiffs += `<span style="color: #ee4231">${part.value}</span>`;
    } else {
      newValueDiffs += part.value;
      oldValueDiffs += part.value;
    }
  });
  return {
    cells: [
      {
        content: [new MediaWikiText(field)],
      },
      {
        content: [new MediaWikiText(oldValueDiffs)],
      },
      {
        content: [new MediaWikiText(newValueDiffs)],
      },
    ],
    minimal: true,
  };
};

/**
 * Build the media wiki content for the differences in two cache index's archives.
 * @param indexDifferences Differences between two cache's index's archives.
 * @param indexFeatures Meta deta for indexes
 * @returns {MediaWikiContent[]}
 */
export const buildIndexDifferences = (
  indexDifferences: IndexDifferences,
  indexFeatures: IndexFeatures
): MediaWikiContent[] => {
  const fileDifferences = flatten(
    Object.values(indexDifferences).map((archive) =>
      Object.values(archive).flat()
    )
  );
  const addedResults: Result[] = _.pluck(fileDifferences, "added").filter(
    (entry) => entry !== null && entry !== undefined
  );

  const removedResults: Result[] = _.pluck(fileDifferences, "removed").filter(
    (entry) => entry !== null && entry !== undefined
  );

  const changedResults: ChangedResult[] = _.pluck(
    fileDifferences,
    "changed"
  ).filter(
    (entry) =>
      entry !== null && entry !== undefined && Object.keys(entry).length > 0
  );

  return buildDiffSection(
    addedResults,
    removedResults,
    changedResults,
    indexFeatures
  );
};

/**
 * Build the media wiki content for the differences in two cache archive files.
 * @param archiveDifferences Differences between two cache's archives
 * @param indexFeatures Meta deta for indexes
 * @returns {MediaWikiContent[]}
 */
export const buildArchiveDifferences = (
  archiveDifferences: ArchiveDifferences,
  indexFeatures: IndexFeatures
): MediaWikiContent[] => {
  const addedResults: Result[] = _.pluck(
    Object.values(archiveDifferences),
    "added"
  ).filter((entry) => entry !== null && entry !== undefined);

  const removedResults: Result[] = _.pluck(
    Object.values(archiveDifferences),
    "removed"
  ).filter((entry) => entry !== null && entry !== undefined);

  const changedResults: ChangedResult[] = _.pluck(
    Object.values(archiveDifferences),
    "changed"
  ).filter(
    (entry) =>
      entry !== null && entry !== undefined && Object.keys(entry).length > 0
  );

  return buildDiffSection(
    addedResults,
    removedResults,
    changedResults,
    indexFeatures
  );
};

/**
 * Builds a section of media wiki content for the differences between two cache files.
 *  This includes the added, removed, and changed results.
 * @param addedResults The added results from the differences
 * @param removedResults The removed results from the differences
 * @param changedResults The changed results from the differences
 * @param indexFeatures Meta data for indexes
 * @returns {MediaWikiContent[]}
 */
export const buildDiffSection = (
  addedResults: Result[],
  removedResults: Result[],
  changedResults: ChangedResult[],
  indexFeatures: IndexFeatures
): MediaWikiContent[] => {
  const content: MediaWikiContent[] = [
    new MediaWikiHeader(indexFeatures.name, 2),
    new MediaWikiBreak(),
  ];

  if (addedResults.length > 0) {
    content.push(...buildResultTable(addedResults, indexFeatures, "added"));
  }

  if (removedResults.length > 0) {
    content.push(...buildResultTable(removedResults, indexFeatures, "removed"));
  }

  if (changedResults.length > 0) {
    content.push(...buildChangedResultTable(changedResults, indexFeatures));
  }
  return content;
};

/**
 * Builds media wiki content for archive file differences.
 *  These tables include all changed fields.
 * @param archiveDifferences Differences between two cache's archives
 * @param indexFeatures Meta deta for indexes
 * @returns {MediaWikiContent[]}
 */
export const buildChangedResultTable = (
  changedResults: ChangedResult[],
  indexFeatures: IndexFeatures
) => {
  const differenceName = resultNameMap.changed;
  const content: MediaWikiContent[] = [];

  const rows: MediaWikiTableRow[] =
    changedResults?.length > 0
      ? changedResults
          .map<MediaWikiTableRow[]>((entry) => {
            const diffKeys = Object.keys(entry).filter((key) => {
              const isIdentifier = (
                indexFeatures.identifiers as string[]
              ).includes(key);
              return (
                !isIdentifier || entry[key].oldValue !== entry[key].newValue
              );
            });
            return diffKeys.length > 0
              ? [
                  {
                    cells: indexFeatures.identifiers
                      .map<MediaWikiTableCell>((identifier) =>
                        entry[identifier]
                          ? {
                              content: formatEntryIdentifier(
                                identifier,
                                entry[identifier].newValue,
                                indexFeatures.urls
                              ),
                              options: {
                                rowspan: diffKeys.length + 1,
                              },
                            }
                          : undefined
                      )
                      .filter((value) => value !== undefined),
                  },
                  ...diffKeys.map<MediaWikiTableRow>((field) =>
                    getFieldDifferencesRow(entry, field)
                  ),
                ]
              : undefined;
          })
          .flat()
          .filter((value) => value !== undefined)
      : [];

  content.push(
    new MediaWikiHeader(`${differenceName} ${indexFeatures.name}`, 3),
    new MediaWikiBreak(),
    new MediaWikiTable({
      rows: [
        {
          header: true,
          minimal: true,
          cells: [
            {
              content: [
                new MediaWikiText(`${differenceName} ${indexFeatures.name}`),
              ],
              options: {
                colspan: 2 + indexFeatures.identifiers.length,
              },
            },
          ],
        },
        {
          header: true,
          minimal: true,
          cells: [
            ...indexFeatures.identifiers.map((identifier) =>
              capitalize(identifier)
            ),
            "Key",
            "Previous Value",
            "New Value",
          ].map((column, index) => ({
            content: [new MediaWikiText(column)],
            options: index == 0 ? { style: "width: 15em" } : undefined,
          })),
        },
        ...rows,
      ],
      options: {
        class: `wikitable sortable sticky-header mw-collapsible${
          rows.length > 50 ? " mw-collapsed" : ""
        }`,
      },
    }),
    new MediaWikiBreak()
  );
  return content;
};

/**
 * Build media wiki content for archive file's additions or removals.
 *  These tables include specified fields from indexFeatures.
 * @param archiveDifferences Differences between two cache's archive files
 * @param indexFeatures Meta deta for indexes
 * @param type Definition for added or removed content
 * @returns
 */
export const buildResultTable = (
  results: Result[],
  indexFeatures: IndexFeatures,
  type: Difference
): MediaWikiContent[] => {
  const differenceName = resultNameMap[type];
  const tableFields = indexFeatures.fields
    .map((field) => field.toString())
    .filter((field) => !!field);
  const fields = [...indexFeatures.identifiers, ...tableFields];
  const content: MediaWikiContent[] = [];

  const rows: MediaWikiTableRow[] =
    results?.length > 0
      ? results.map((entry) => {
          const identifierCells = indexFeatures.identifiers
            .filter((identifier) => !!identifier)
            .map((identifier) => ({
              content: formatEntryIdentifier(
                identifier,
                entry[identifier],
                indexFeatures.urls
              ),
            }));
          return {
            cells: [
              ...identifierCells,
              ...tableFields.map((field) => ({
                content: [
                  new MediaWikiText(formatEntryValue(field, entry[field])),
                ],
              })),
            ],
            minimal: true,
          };
        })
      : [];

  content.push(
    new MediaWikiHeader(`${differenceName} ${indexFeatures.name}`, 3),
    new MediaWikiBreak(),
    new MediaWikiTable({
      rows: [
        {
          header: true,
          minimal: true,
          cells: [
            {
              content: [
                new MediaWikiText(`${differenceName} ${indexFeatures.name}`),
              ],
              options: {
                colspan: fields.length,
              },
            },
          ],
        },
        {
          header: true,
          minimal: true,
          cells: fields
            .filter((field) => !!field)
            .map((field, index) => ({
              content: [new MediaWikiText(field)],
              options: index == 0 ? { style: "width: 15em" } : undefined,
            })),
        },
        ...rows,
      ],
      options: {
        class: `wikitable sortable sticky-header mw-collapsible${
          rows.length > 50 ? " mw-collapsed" : ""
        }`,
      },
    }),
    new MediaWikiBreak()
  );
  return content;
};
