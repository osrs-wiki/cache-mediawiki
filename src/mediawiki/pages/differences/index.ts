export { default as differencesPageBuilder } from "./differences";
export * from "./differences.types";
