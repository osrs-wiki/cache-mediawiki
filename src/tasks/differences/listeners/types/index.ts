import { areaListener } from "./areas";
import { itemListener } from "./items";
import { musicListener } from "./music";
import { npcListener } from "./npcs";
import { objectListener } from "./objects";
import { CacheChangeListener } from "../listeners.types";

export const cacheListeners: CacheChangeListener[] = [
  areaListener,
  itemListener,
  musicListener,
  npcListener,
  objectListener,
];
