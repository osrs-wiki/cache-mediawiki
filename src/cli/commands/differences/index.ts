import differences from "./differences";

export default differences;
