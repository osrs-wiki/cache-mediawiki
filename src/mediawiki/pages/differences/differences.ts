import { MediaWikiBuilder, MediaWikiTOC } from "@osrs-wiki/mediawiki-builder";
import _ from "underscore";

import { IndexFeatures, indexNameMap } from "./differences.types";
import {
  buildArchiveDifferences,
  buildIndexDifferences,
} from "./differences.utils";

import { CacheDifferences } from "@/tasks/differences/differences.types";
import { IndexType } from "@/utils/cache2/types";

/**
 * Retrieve a MediaWikiBuilder filled with content from two cache differences.
 * @param differences The differences between two caches.
 * @returns {MediaWikiBuilder}
 */
const differencesPageBuilder = (
  differences: CacheDifferences
): MediaWikiBuilder => {
  const builder = new MediaWikiBuilder();

  builder.addContents([new MediaWikiTOC()]);

  // TODO: Build from supported index and archives
  /*Object.keys(indexNameMap).forEach((index) => {
    const indexFeatureMap = indexNameMap[index as unknown as IndexType];
    if (indexFeatureMap) {
      if ("name" in indexFeatureMap) {
        builder.addContents(
          buildArchiveDifferences(differences[index as unknown as IndexType]?.[], indexFeatureMap)
        );
      } else {
        Object.keys(indexFeatureMap).forEach((archive) => {
          const archiveFeatureMap =
            indexFeatureMap[archive as unknown as number];
        });
      }
    }
  });*/

  Object.keys(differences).forEach((index) => {
    const indexFeatureMap = indexNameMap[index as unknown as IndexType];
    if (indexFeatureMap) {
      const indexDifferences = differences[index as unknown as number];
      if ("name" in indexFeatureMap) {
        const indexFeature = indexFeatureMap as IndexFeatures;
        if (Object.keys(indexDifferences).length > 0) {
          builder.addContents(
            buildIndexDifferences(indexDifferences, indexFeature)
          );
        }
      } else {
        Object.keys(indexDifferences).forEach((archive) => {
          const archiveNumber = archive as unknown as number;
          const archiveDifferences = indexDifferences[archiveNumber];
          const indexFeature = indexFeatureMap[archiveNumber];
          if (indexFeature && Object.keys(archiveDifferences).length > 0) {
            builder.addContents(
              buildArchiveDifferences(archiveDifferences, indexFeature)
            );
          }
        });
      }
    }
  });
  return builder;
};

export default differencesPageBuilder;
