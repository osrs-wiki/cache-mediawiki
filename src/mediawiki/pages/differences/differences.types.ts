import { Difference } from "../../../tasks/differences/differences.types";

import {
  Animation,
  Area,
  ConfigType,
  DBRow,
  Enum,
  IndexType,
  Item,
  NPC,
  Obj,
  Param,
  Region,
  SpotAnim,
  Sprites,
  Struct,
  VarPlayer,
  Varbit,
  Widget,
} from "@/utils/cache2";
import { Loadable } from "@/utils/cache2/Loadable";
import { regionToWorldMapURL } from "@/utils/url-generation";

export type IndexFeature<T extends Loadable, Name> = {
  name: Name;
  identifiers: (keyof T)[];
  fields: (keyof T)[];
  urls?: IndexFieldURLs;
};

export type IndexURLType = "abex" | "chisel" | "world";

// URL Generator function signature
export type URLGeneratorFunction = (
  value: unknown,
  context?: URLGeneratorContext
) => string;

// Context provided to URL generator functions
export type URLGeneratorContext = {
  fieldName: string;
  entityType: string;
  allFields?: Record<string, unknown>;
};

// Template URL with placeholder substitution
export type TemplateURL = string; // Contains {fieldName} placeholders

// URL definition for a specific provider
export type URLDefinition = TemplateURL | URLGeneratorFunction;

// Field-specific URL mapping
export type FieldURLs = {
  [provider in IndexURLType]?: URLDefinition;
};

// New field-specific URL structure
export type IndexFieldURLs = {
  [fieldName: string]: FieldURLs;
};

export type IndexFeatures =
  | IndexFeature<Animation, "Animations">
  | IndexFeature<Area, "Areas">
  | IndexFeature<DBRow, "Database Rows">
  | IndexFeature<Enum, "Enums">
  | IndexFeature<Item, "Items">
  | IndexFeature<NPC, "Npcs">
  | IndexFeature<Obj, "Objects">
  | IndexFeature<Param, "Params">
  | IndexFeature<Region, "Regions">
  | IndexFeature<SpotAnim, "Spot Anims">
  | IndexFeature<Sprites, "Sprites">
  | IndexFeature<Struct, "Structs">
  | IndexFeature<Varbit, "Varbits">
  | IndexFeature<VarPlayer, "VarPlayers">
  | IndexFeature<Widget, "Widgets">;

export const resultNameMap: { [key in Difference]: string } = {
  added: "New",
  changed: "Diff",
  removed: "Removed",
};

export const indexNameMap: {
  [key in IndexType]?: { [key: number]: IndexFeatures } | IndexFeatures;
} = {
  [IndexType.Configs]: {
    [ConfigType.Sequence]: {
      name: "Animations",
      identifiers: ["id", "gameVal"],
      fields: ["debugName", "leftHandItem", "rightHandItem"],
      urls: {},
    },
    [ConfigType.Area]: {
      name: "Areas",
      identifiers: ["name", "id"],
      fields: ["category", "spriteId"],
      urls: {},
    },
    [ConfigType.Object]: {
      name: "Objects",
      identifiers: ["name", "id", "gameVal"],
      fields: ["actions"],
      urls: {
        id: {
          chisel: "https://chisel.weirdgloop.org/moid/object_id.html#{id}",
          abex: "https://abextm.github.io/cache2/#/viewer/obj/{id}",
        },
      },
    },
    [ConfigType.Enum]: {
      name: "Enums",
      identifiers: ["id"],
      fields: ["defaultValue", "map"],
      urls: {
        id: {
          chisel:
            "https://chisel.weirdgloop.org/structs/index.html?type=enums&id={id}",
          abex: "https://abextm.github.io/cache2/#/viewer/enum/{id}",
        },
      },
    },
    [ConfigType.Npc]: {
      name: "Npcs",
      identifiers: ["name", "id", "gameVal"],
      fields: ["combatLevel", "actions"],
      urls: {
        id: {
          chisel: "https://chisel.weirdgloop.org/moid/npc_id.html#{id}",
          abex: "https://abextm.github.io/cache2/#/viewer/npc/{id}",
        },
      },
    },
    [ConfigType.Item]: {
      name: "Items",
      identifiers: ["name", "id", "gameVal"],
      fields: [
        "examine",
        "isMembers",
        "isGrandExchangable",
        "isStackable",
        "noteLinkedItem",
        "inventoryActions",
        "placeholderLinkedItem",
        "price",
        "weight",
      ],
      urls: {
        id: {
          chisel: "https://chisel.weirdgloop.org/moid/item_id.html#{id}",
          abex: "https://abextm.github.io/cache2/#/viewer/item/{id}",
        },
      },
    },
    [ConfigType.Params]: {
      name: "Params",
      identifiers: ["id"],
      fields: ["type", "defaultInt", "defaultString", "isMembers"],
      urls: {
        id: {
          chisel:
            "https://chisel.weirdgloop.org/structs/index.html?type=enums&id={id}",
          abex: "https://abextm.github.io/cache2/#/viewer/enum/{id}",
        },
      },
    },
    [ConfigType.Struct]: {
      name: "Structs",
      identifiers: ["id"],
      fields: ["params"],
      urls: {
        id: {
          chisel:
            "https://chisel.weirdgloop.org/structs/index.html?type=structs&id={id}",
          abex: "https://abextm.github.io/cache2/#/viewer/struct/{id}",
        },
      },
    },
    [ConfigType.DbRow]: {
      name: "Database Rows",
      identifiers: ["id", "gameVal"],
      fields: ["table", "values"],
      urls: {
        id: {
          abex: "https://abextm.github.io/cache2/#/viewer/dbrow/{id}",
        },
      },
    },
    [ConfigType.SpotAnim]: {
      name: "Spot Anims",
      identifiers: ["id", "gameVal"],
      fields: ["modelId", "animationId"],
      urls: {},
    },
    [ConfigType.VarBit]: {
      name: "Varbits",
      identifiers: ["id", "gameVal"],
      fields: ["index", "leastSignificantBit", "mostSignificantBit"],
      urls: {
        id: {
          chisel: "https://chisel.weirdgloop.org/varbs/display?varbit={id}",
        },
      },
    },
    [ConfigType.VarPlayer]: {
      name: "VarPlayers",
      identifiers: ["id", "gameVal"],
      fields: [],
      urls: {
        id: {
          chisel: "https://chisel.weirdgloop.org/varbs/display?varplayer={id}",
        },
      },
    },
  },
  [IndexType.Interfaces]: {
    name: "Widgets",
    identifiers: ["id", "gameVal"],
    fields: [
      "parentId",
      "type",
      "name",
      "text",
      "tooltip",
      "actions",
      "configActions",
    ],
    urls: {
      id: {
        abex: "https://abextm.github.io/cache2/#/viewer/interface/{id}",
      },
    },
  },
  [IndexType.Maps]: {
    name: "Regions",
    identifiers: ["id", "name"],
    fields: ["regionX", "regionY"],
    urls: {
      id: {
        world: regionToWorldMapURL,
      },
    },
  },
  [IndexType.Sprites]: {
    name: "Sprites",
    identifiers: ["id", "gameVal"],
    fields: ["width", "height"],
    urls: {
      id: {
        abex: "https://abextm.github.io/cache2/#/viewer/sprite/{id}",
      },
    },
  },
};
