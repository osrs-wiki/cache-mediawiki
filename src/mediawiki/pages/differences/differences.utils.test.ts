import {
  MediaWikiExternalLink,
  MediaWikiText,
  MediaWikiLink,
} from "@osrs-wiki/mediawiki-builder";

import { URLGeneratorFunction } from "./differences.types";
import {
  getFieldDifferencesRow,
  formatEntryValue,
  formatEntryIdentifier,
} from "./differences.utils";

import { jagexHSLtoHex } from "@/utils/colors";
import { regionToWorldMapURL } from "@/utils/url-generation";

// Mock the jagexHSLtoHex function
jest.mock("@/utils/colors", () => ({
  jagexHSLtoHex: jest.fn((color) => `#mock-${color}`),
}));

describe("differences builder utils", () => {
  describe("getFieldDifferencesRow", () => {
    test("one difference", () => {
      expect(
        getFieldDifferencesRow(
          { test: { oldValue: "test removed", newValue: "test added" } },
          "test"
        )
      ).toMatchSnapshot();
    });

    test("multiple difference", () => {
      expect(
        getFieldDifferencesRow(
          {
            test: {
              oldValue: "test removed first",
              newValue: "test added second",
            },
          },
          "test"
        )
      ).toMatchSnapshot();
    });
  });

  describe("formatEntryValue", () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    test("should return empty string for undefined values", () => {
      expect(formatEntryValue("someField", undefined)).toBe("");
    });

    test("should return empty string for empty objects", () => {
      expect(formatEntryValue("someField", {})).toBe("");
    });

    test("should format color fields with HTML span", () => {
      const colorValue = 12345;
      const result = formatEntryValue("color", colorValue);

      expect(jagexHSLtoHex).toHaveBeenCalledWith(colorValue);
      expect(result).toBe(
        `<span style = "background-color: #mock-${colorValue};color:white">${colorValue}</span>`
      );
    });

    test("should format color fields with HTML span when field has 'color' in name", () => {
      const colorValue = 54321;
      const result = formatEntryValue("textColor", colorValue);

      expect(jagexHSLtoHex).toHaveBeenCalledWith(colorValue);
      expect(result).toBe(
        `<span style = "background-color: #mock-${colorValue};color:white">${colorValue}</span>`
      );
    });

    test("should format arrays by converting to comma-separated string", () => {
      const arrayValue = [1, 2, 3];
      expect(formatEntryValue("items", arrayValue)).toBe("[1, 2, 3]");
    });

    test("should replace null/undefined values in arrays with 'None'", () => {
      const arrayWithNulls = [1, null, 3, undefined];
      expect(formatEntryValue("items", arrayWithNulls)).toBe(
        "[1, None, 3, None]"
      );
    });

    test("should return empty string for empty arrays", () => {
      expect(formatEntryValue("emptyArray", [])).toBe("");
    });

    test("should return value as string for string values", () => {
      expect(formatEntryValue("name", "Test String")).toBe("Test String");
    });

    test("should return value as string for numeric values", () => {
      expect(formatEntryValue("count", 42)).toBe("42");
    });

    test("should format objects by converting to string with custom formatting", () => {
      const objectValue = { key1: "value1", key2: 2 };
      expect(formatEntryValue("config", objectValue)).toBe(
        "{'key1': 'value1', 'key2': 2}"
      );
    });

    test("should handle complex nested objects", () => {
      const complexObject = { a: { b: 1 }, c: [2, 3] };
      expect(formatEntryValue("complex", complexObject)).toBe(
        "{'a': {'b': 1}, 'c': [2, 3]}"
      );
    });
  });

  describe("formatEntryIdentifier", () => {
    test("should strip HTML tags from name identifiers", () => {
      const result = formatEntryIdentifier(
        "name",
        "<col=00ffff>Tornado</col>",
        {}
      );
      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiLink);
      // The function should create a MediaWikiLink with stripped tags
      const built = result[0].build();
      expect(built).toContain("Tornado");
      expect(built).not.toContain("<col=");
      expect(built).not.toContain("</col>");
    });

    test("should handle empty name values", () => {
      const result = formatEntryIdentifier("name", "", {});
      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiText);
    });

    test("should handle null name values", () => {
      const result = formatEntryIdentifier("name", null, {});
      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiText);
    });

    test("should handle field-specific URL format with templates", () => {
      const fieldSpecificURLs = {
        id: {
          chisel:
            "https://chisel.weirdgloop.org/moid/npc_id.html#{id}" as const,
          abex: "https://abextm.github.io/cache2/#/viewer/npc/{id}" as const,
        },
      };

      const result = formatEntryIdentifier("id", 123, fieldSpecificURLs);
      expect(result).toHaveLength(2);
      expect(result[0]).toBeInstanceOf(MediaWikiExternalLink);
      expect(result[1]).toBeInstanceOf(MediaWikiExternalLink);

      // Verify URLs are generated correctly by building the content
      const firstBuilt = result[0].build();
      const secondBuilt = result[1].build();
      expect(firstBuilt).toContain(
        "https://chisel.weirdgloop.org/moid/npc_id.html#123"
      );
      expect(secondBuilt).toContain(
        "https://abextm.github.io/cache2/#/viewer/npc/123"
      );
    });

    test("should handle function-based URLs", () => {
      const functionURLs = {
        id: {
          world: regionToWorldMapURL,
        },
      };

      const regionId = (50 << 8) | 30;
      const result = formatEntryIdentifier("id", regionId, functionURLs);
      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiExternalLink);

      const built = result[0].build();
      expect(built).toContain("osrs.world");
      expect(built).toContain("cx=3200"); // 50 << 6
      expect(built).toContain("cz=1920"); // 30 << 6
    });

    test("should fall back to text for fields without URLs", () => {
      const fieldSpecificURLs = {
        id: { chisel: "https://example.com/{id}" as const },
      };

      const result = formatEntryIdentifier(
        "unsupported",
        "value",
        fieldSpecificURLs
      );
      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiText);
    });

    test("should handle empty URL configuration", () => {
      const result = formatEntryIdentifier("id", 123, {});
      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiText);
    });

    test("should pass context to URL generation", () => {
      const mockFunction = jest.fn().mockReturnValue("https://test.com");
      const functionURLs = {
        id: {
          world: mockFunction as unknown as typeof regionToWorldMapURL,
        },
      };

      const allFields = { name: "Test Item", id: 123 };
      const result = formatEntryIdentifier(
        "id",
        123,
        functionURLs,
        allFields,
        "npc"
      );

      expect(mockFunction).toHaveBeenCalledWith(123, {
        fieldName: "id",
        entityType: "npc",
        allFields,
      });
      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiExternalLink);
    });

    test("should handle URL generation errors gracefully", () => {
      const errorFunction = jest.fn().mockImplementation(() => {
        throw new Error("URL generation failed");
      });

      const consoleWarnSpy = jest.spyOn(console, "warn").mockImplementation();

      const functionURLs = {
        id: {
          world: errorFunction as unknown as typeof regionToWorldMapURL,
        },
      };

      const result = formatEntryIdentifier("id", 123, functionURLs);

      expect(result).toHaveLength(1);
      expect(result[0]).toBeInstanceOf(MediaWikiText);
      expect(consoleWarnSpy).toHaveBeenCalledWith(
        "Failed to generate URL for id:world",
        expect.any(Error)
      );

      consoleWarnSpy.mockRestore();
    });

    test("should create external links with correct content", () => {
      const fieldSpecificURLs = {
        id: {
          chisel: "https://first.com/{id}" as const,
          abex: "https://second.com/{id}" as const,
        },
      };

      const result = formatEntryIdentifier("id", 456, fieldSpecificURLs);
      expect(result).toHaveLength(2);

      // Check that the links are built correctly
      const firstBuilt = result[0].build();
      const secondBuilt = result[1].build();

      expect(firstBuilt).toContain("[https://first.com/456 456]");
      expect(secondBuilt).toContain("[https://second.com/456 (1)]");
    });
  });
});
