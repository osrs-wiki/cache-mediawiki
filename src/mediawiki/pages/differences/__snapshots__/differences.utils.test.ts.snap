// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`differences builder utils getFieldDifferencesRow multiple difference 1`] = `
Object {
  "cells": Array [
    Object {
      "content": Array [
        MediaWikiText {
          "children": "test",
          "styling": undefined,
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": "test <span style=\\"color: #ee4231\\">removed first</span>",
          "styling": undefined,
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": "test <span style=\\"color: #6bc71f\\">added second</span>",
          "styling": undefined,
        },
      ],
    },
  ],
  "minimal": true,
}
`;

exports[`differences builder utils getFieldDifferencesRow one difference 1`] = `
Object {
  "cells": Array [
    Object {
      "content": Array [
        MediaWikiText {
          "children": "test",
          "styling": undefined,
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": "test <span style=\\"color: #ee4231\\">removed</span>",
          "styling": undefined,
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": "test <span style=\\"color: #6bc71f\\">added</span>",
          "styling": undefined,
        },
      ],
    },
  ],
  "minimal": true,
}
`;
